performance(jdbc): add deleted field condition when allowDeleted=true

lw-yang opened the pull request:
### What changes are being made and why?

close #12179 
---

### How the changes have been QAed?

exist ut
---

lw-yang committed:
fix(jdbc): add deleted field condition when allowDeleted=true

loicmathieu commented:
It's surprising that the optimizer didn't use the index if the field is missing and use it if we provide all possible values of the field. Maybe it's a limitation of the MySQL optimizer.

Can you give some numbers before/after your change?

lw-yang commented:
sure, below are the test cases, the table contains over 2 million records.

When the SQL query ​​excludes the `deleted` condition​​, it takes ​​63 seconds​​ and performs a full table scan.
When the query ​​includes `deleted`​​, it takes ​​8 ms​​ and uses the composite index efficiently.

```sql
select count(*) from `executions`;
```
<img width="924" height="180" alt="image" src="https://github.com/user-attachments/assets/f50ad7e0-3915-4b5c-9088-2b2371637a05" />

```sql
select * from `executions` where tenant_id is null and deleted in (true,false) and namespace = 'dt.10007';
```
elapsed time: 8ms​​

```sql
select * from `executions` where tenant_id is null  and namespace = 'dt.10007';
```
elapsed time: 63s​​

```sql
explain select * from `executions` where tenant_id is null and namespace = 'dt.15650';
```
<img width="2040" height="178" alt="image" src="https://github.com/user-attachments/assets/58254747-a69c-48e4-a4dc-b1ffdb1910c4" />


```sql
explain select * from `executions` where tenant_id is null and deleted in (true,false) and namespace = 'dt.15650';
```
<img width="3090" height="184" alt="image" src="https://github.com/user-attachments/assets/43b8a8f6-2e05-49d6-80f7-c3da09b4e88d" />

---
the introduction to multiple-column-indexes from the official MySQL documentation.
https://dev.mysql.com/doc/refman/8.0/en/multiple-column-indexes.html
<img width="2424" height="870" alt="image" src="https://github.com/user-attachments/assets/d2aab8b8-5a1c-42ff-9801-24e790161c6c" />

loicmathieu reviewed (changes requested):
Overall, it's OK for me.

Can you add a comment, something like:

> Always include deleted in the query filters as most database optimizers can only use and index if the leftmost columns are used in the query

This would prevent someone to remove this later ;)

I need to check in Postgres if it didn't cause any issue before merging it.
Can you also tell me if you want it to be backported to a specific version, as this is a small change, I can backport it even if we usually don't backport performance improvements.

lw-yang committed:
fix(jdbc): add comment

lw-yang commented:
I’ve added the comment, we don't need to backport

loicmathieu reviewed (approved):
LGTM, thanks

loicmathieu commented:
Tested it with Postgres and the explain plan seems little better with the new clause :tada:
